File: 740. 删除与获得点数/740.py
class Solution:
    def deleteAndEarn(self, nums: List[int]) -> int:
        if(len(nums)==0):
            return 0
        elif(len(nums)==1):
            return nums[0]

        length = max(nums)+1
        all_count = [0]*length

        dp = [0]*length
        for i in nums:
            all_count[i] += 1

        dp[1] = all_count[1]
        for i in range(2,length):
            dp[i] = max(dp[i-1],dp[i-2]+all_count[i]*i)
        # print(dp)
        return dp[-1]
        

File: 740. 删除与获得点数/740.js
/**
 * @param {number[]} nums
 * @return {number}
 */
 var deleteAndEarn = function(nums) {
    if(nums.length==0){
        return 0;
    }
    else if(nums.length==1){
        return nums[0];
    }
    let length = Math.max(...nums);
    const all_nums = new Array(length+1).fill(0);
    const dp = new Array(length+1).fill(0);
    for(const i of nums){
        all_nums[i] ++;
    }
    dp[0] = 0;
    dp[1] = all_nums[1];
    for(let i=2;i<length+1;i++){
        dp[i] = Math.max(dp[i-2]+all_nums[i]*i,dp[i-1]);
    }
    return dp[length];
};